subcommands/keys/ca_revoke_device_ca.go

```diff
@@ -0,0 +1,224 @@
+		ca := x509.LoadCertFromFile(filename)
+		toRevoke[ca.SerialNumber.Text(10)] = crlReason
+	}
+
+	caList, err := api.FactoryGetCA(factory)
+	subcommands.DieNotNil(err)
+	validSerials := make(map[string]bool, 0)
+	for len(caList.CaCrt) > 0 {
+		block, remaining := pem.Decode([]byte(caList.CaCrt))
+		if block == nil {
+			// could be excessive whitespace
+			caList.CaCrt = strings.TrimSpace(string(remaining))
+			continue
+		}
+		caList.CaCrt = string(remaining)
+		c, err := x509Lib.ParseCertificate(block.Bytes)
+		if err != nil {
+			fmt.Println("WARNING: Failed to parse certificate from server:" + err.Error())
+			continue
+		}
+		validSerials[c.SerialNumber.Text(10)] = true
+	}
+	for serial := range toRevoke {
+		if _, ok := validSerials[serial]; !ok {
+			subcommands.DieNotNil(fmt.Errorf("There is no actual device CA with serial %s", serial))
+		}
+	}
+
+	fmt.Println("Signing CRL by factory root CA")
+	certs := client.CaCerts{CaRevokeCrl: x509.CreateCrl(toRevoke)}
+
+	if dryRun {
+		fmt.Println(certs.CaRevokeCrl)
+		if pretty {
+			prettyPrintCrl(certs.CaRevokeCrl)
+		}
+		return
+	}
+
+	fmt.Println("Uploading CRL to Foundries.io")
+	subcommands.DieNotNil(api.FactoryPatchCA(factory, certs))
+}
+
+func prettyPrintCrl(crlPem string) {
+	block, remaining := pem.Decode([]byte(crlPem))
+	if block == nil || len(remaining) > 0 {
+		subcommands.DieNotNil(errors.New("Invalid PEM block"), "Failed to parse generated CRL:")
+		return // linter
+	}
+	c, err := x509Lib.ParseRevocationList(block.Bytes)
+	subcommands.DieNotNil(err, "Failed to parse generated CRL:")
+	fmt.Println("Certificate Revocation List:")
+	fmt.Println("\tIssuer:", c.Issuer)
+	fmt.Println("\tValidity:")
+	fmt.Println("\t\tNot Before:", c.ThisUpdate)
+	fmt.Println("\t\tNot After:", c.NextUpdate)
+	fmt.Println("\tSignature Algorithm:", c.SignatureAlgorithm)
+	fmt.Println("\tSignature:", hex.EncodeToString(c.Signature))
+	fmt.Println("\tRevoked Certificates:")
+	for _, crt := range c.RevokedCertificates {
+		fmt.Println("\t\tSerial:", crt.SerialNumber.Text(10))
+		fmt.Println("\t\t\tRevocation Date:", crt.RevocationTime)
+		if len(crt.Extensions) > 0 {
+			fmt.Println("\t\t\tExtensions:")
+			for _, ext := range crt.Extensions {
+				if ext.Id.String() == "2.5.29.21" {
+					fmt.Print("\t\t\t\tx509v3 Reason Code:")
+					if ext.Critical {
+						fmt.Print("(critical)")
+					}
+					var val asn1.Enumerated
+					if _, err := asn1.Unmarshal(ext.Value, &val); err != nil {
+						fmt.Println(err)
+					} else {
+						readable := map[int]string{
+							x509.CrlCaRevoke:  "Revoke",
+							x509.CrlCaDisable: "Disable",
+							x509.CrlCaRenew:   "Renew",
+						}[int(val)]
+						fmt.Println("\n\t\t\t\t\t", readable, "-", val)
+					}
+				} else {
+					fmt.Println("\t\t\t\tUnknown OID", ext.Id.String())
+				}
+			}
+		}
+	}
+	if len(c.Extensions) > 0 {
+		fmt.Println("\tExtensions:")
+		for _, ext := range c.Extensions {
+			if ext.Id.String() == "2.5.29.35" {
```

Comment:
Why have we a fixed id?
Could we add it to a const and then add a comment on top to clarify?

Reply:
I followed the same pattern as `ca_show` does.
These are extension OIDs we recognize (because certs we generate have them).
Their meaning must be evident from the code; otherwise we have a problem.
The constant at the top would disperse the (cryptic enough) code, make it less clear.

Reply:
Oh. I see those values are the specific OID.
They are standard definitions. 
Thank you. Yep makes sense.